File: app/src/main/java/net/ivpn/client/ui/split/SplitTunnelingRecyclerViewAdapter.java
package net.ivpn.client.ui.split;

/*
 IVPN Android app
 https://github.com/ivpn/android-app
 <p>
 Created by Oleksandr Mykhailenko.
 Copyright (c) 2020 Privatus Limited.
 <p>
 This file is part of the IVPN Android app.
 <p>
 The IVPN Android app is free software: you can redistribute it and/or
 modify it under the terms of the GNU General Public License as published by the Free
 Software Foundation, either version 3 of the License, or (at your option) any later version.
 <p>
 The IVPN Android app is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.
 <p>
 You should have received a copy of the GNU General Public License
 along with the IVPN Android app. If not, see <https://www.gnu.org/licenses/>.
*/

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;

import net.ivpn.client.R;
import net.ivpn.client.databinding.ApplicationItemBinding;
import net.ivpn.client.ui.split.data.ApplicationItem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

public class SplitTunnelingRecyclerViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int APP_ITEM = 0;
    private static final int DESCRIPTION_ITEM = 1;

    private List<ApplicationItem> allApps = new LinkedList<>();
    private Set<String> disallowedApps = new HashSet<>();
    private List<ApplicationItemBinding> bindings = new LinkedList<>();
    private OnApplicationItemSelectionChangedListener listener;

    @Inject
    SplitTunnelingRecyclerViewAdapter() {
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return DESCRIPTION_ITEM;
        }
        return APP_ITEM;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        switch (viewType) {
            case APP_ITEM: {
                ApplicationItemBinding binding = ApplicationItemBinding.inflate(layoutInflater, parent, false);
                bindings.add(binding);
                return new ApplicationInfoViewHolder(binding);
            }
            default: {
                return new DescriptionViewHolder(layoutInflater.inflate(R.layout.description_item, parent, false));
            }
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof ApplicationInfoViewHolder) {
            ((ApplicationInfoViewHolder) holder).bind(allApps.get(position - 1));
        }
    }

    @Override
    public int getItemCount() {
        return allApps.size() + 1;
    }

    public void setSelectionChangedListener(OnApplicationItemSelectionChangedListener listener) {
        this.listener = listener;
    }

    MenuHandler getMenuHandler() {
        return new MenuHandler();
    }

    public void setApplicationsInfoList(List<ApplicationItem> infoList) {
        this.allApps = new ArrayList<>(infoList);
        Collections.sort(allApps, ApplicationItem.comparator);
        notifyDataSetChanged();
    }

    public void setDisallowedApps(List<String> disallowedApps) {
        this.disallowedApps = new HashSet<>(disallowedApps);
        if (listener != null) {
            listener.onItemsSelectionStateChanged(disallowedApps.size() == 0);
        }
    }

    private void selectAll() {
        disallowedApps = new HashSet<>();
        listener.onItemsSelectionStateChanged(true);
        //Better alternative than notifyItemRangeChanged for our case;
        setSelections(true);
    }

    private void deselectAll() {
        disallowedApps = new HashSet<>();
        for (ApplicationItem app : allApps) {
            disallowedApps.add(app.getPackageName());
        }
        listener.onItemsSelectionStateChanged(false);
        //Better alternative than notifyItemRangeChanged for our case;
        setSelections(false);
    }

    private void setSelections(boolean isSelected) {
        for (ApplicationItemBinding binding : bindings) {
            binding.checkbox.setChecked(isSelected);
        }
    }

    class ApplicationInfoViewHolder extends RecyclerView.ViewHolder
            implements CompoundButton.OnCheckedChangeListener, View.OnClickListener {

        private ApplicationItemBinding binding;
        private ApplicationItem applicationItem;

        ApplicationInfoViewHolder(ApplicationItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(ApplicationItem applicationItem) {
            this.applicationItem = applicationItem;
            binding.setApplication(applicationItem);
            boolean isChecked = !disallowedApps.contains(applicationItem.getPackageName());
            binding.checkbox.setChecked(isChecked);
            binding.checkbox.setOnCheckedChangeListener(this);
            binding.contentLayout.setOnClickListener(this);
            binding.executePendingBindings();
        }

        @Override
        public void onCheckedChanged(CompoundButton compoundButton, boolean isSelected) {
            if (isSelected) {
                disallowedApps.remove(applicationItem.getPackageName());
            } else {
                disallowedApps.add(applicationItem.getPackageName());
            }
            listener.onItemsSelectionStateChanged(disallowedApps.size() == 0);
            listener.onApplicationItemSelectionChanged(applicationItem, isSelected);
        }

        @Override
        public void onClick(View view) {
            boolean isNotAllowed = disallowedApps.contains(applicationItem.getPackageName());
            binding.checkbox.setChecked(isNotAllowed);
        }
    }

    class DescriptionViewHolder extends RecyclerView.ViewHolder {

        DescriptionViewHolder(View itemView) {
            super(itemView);
        }
    }

    class MenuHandler {
        void selectAll() {
            SplitTunnelingRecyclerViewAdapter.this.selectAll();
        }

        void deselectAll() {
            SplitTunnelingRecyclerViewAdapter.this.deselectAll();
        }
    }
}

File: core/src/main/java/net/ivpn/core/v2/splittunneling/SplitTunnelingRecyclerViewAdapter.kt
package net.ivpn.core.v2.splittunneling

/*
 IVPN Android app
 https://github.com/ivpn/android-app

 Created by Oleksandr Mykhailenko.
 Copyright (c) 2023 IVPN Limited.

 This file is part of the IVPN Android app.

 The IVPN Android app is free software: you can redistribute it and/or
 modify it under the terms of the GNU General Public License as published by the Free
 Software Foundation, either version 3 of the License, or (at your option) any later version.

 The IVPN Android app is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.
 You should have received a copy of the GNU General Public License
 along with the IVPN Android app. If not, see <https://www.gnu.org/licenses/>.
*/

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import net.ivpn.core.R
import net.ivpn.core.databinding.ApplicationItemBinding
import net.ivpn.core.databinding.AppsSearchItemBinding
import net.ivpn.core.v2.splittunneling.holder.ApplicationInfoViewHolder
import net.ivpn.core.v2.splittunneling.holder.AppsSearchViewHolder
import net.ivpn.core.v2.splittunneling.items.ApplicationItem
import net.ivpn.core.v2.splittunneling.items.SearchItem
import net.ivpn.core.v2.splittunneling.items.SplitTunnelingItem
import java.util.*
import javax.inject.Inject

class SplitTunnelingRecyclerViewAdapter @Inject internal constructor()
    : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var allApps: ArrayList<ApplicationItem> = ArrayList()
    private var disallowedApps: MutableSet<String> = HashSet()

    private val bindings: MutableList<ApplicationItemBinding> = LinkedList()
    private var searchBinding: AppsSearchItemBinding? = null

    private var listener: OnApplicationItemSelectionChangedListener? = null
    private val itemListener = getApplicationItemActionListener()

    private var isFiltering = false
    private var filteredApps = arrayListOf<SplitTunnelingItem>()

    override fun getItemViewType(position: Int): Int {
        return when (position) {
            0 -> SEARCH_ITEM
            else -> APP_ITEM
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        return when (viewType) {
            SEARCH_ITEM -> {
                val binding = AppsSearchItemBinding.inflate(layoutInflater, parent, false)
                AppsSearchViewHolder(binding, searchFilter)
            }
            APP_ITEM -> {
                val binding = ApplicationItemBinding.inflate(layoutInflater, parent, false)
                bindings.add(binding)
                ApplicationInfoViewHolder(binding, itemListener)
            }
            else -> {
                DescriptionViewHolder(layoutInflater.inflate(R.layout.description_item, parent, false))
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is ApplicationInfoViewHolder) {
            val app = filteredApps[position]
            if (app is ApplicationItem) {
                holder.bind(app)
            }
        } else if (holder is AppsSearchViewHolder) {
            searchBinding = holder.binding
        }
    }

    override fun getItemCount(): Int {
        return filteredApps.size
    }

    fun setSelectionChangedListener(listener: OnApplicationItemSelectionChangedListener?) {
        this.listener = listener
    }

    val menuHandler: MenuHandler
        get() = MenuHandler()

    fun setApps(allApps: ArrayList<ApplicationItem>, disallowedApps: ArrayList<String>) {
        this.allApps = ArrayList(allApps)
        this.disallowedApps = HashSet(disallowedApps)
        for (app in allApps) {
            app.isAllowed = !disallowedApps.contains(app.packageName)
        }
        Collections.sort(this.allApps, ApplicationItem.comparator)
        searchBinding?.search?.let {
            searchFilter.filter(it.query)
        } ?: run {
            filteredApps = prepareDataToShow(this.allApps)
            notifyDataSetChanged()
        }
        listener?.onItemsSelectionStateChanged(disallowedApps.isEmpty())
    }

    private fun getApplicationItemActionListener(): OnApplicationItemAction {
        return object : OnApplicationItemAction {
            override fun onApplicationStateChanged(item: ApplicationItem, isAllowed: Boolean) {
                if (isAllowed) {
                    disallowedApps.remove(item.packageName)
                } else {
                    disallowedApps.add(item.packageName)
                }
                listener?.onItemsSelectionStateChanged(disallowedApps.size == 0)
                listener?.onApplicationItemSelectionChanged(item, isAllowed)
            }
        }
    }

    private fun selectAll() {
        disallowedApps = HashSet()
        for (app in allApps) {
            app.isAllowed = true
        }
        listener?.onItemsSelectionStateChanged(true)
        //Better alternative than notifyItemRangeChanged for our case;
        setSelections(true)
    }

    private fun deselectAll() {
        disallowedApps = HashSet()
        for (app in allApps) {
            app.isAllowed = false
            disallowedApps.add(app.packageName)
        }
        listener?.onItemsSelectionStateChanged(false)
        //Better alternative than notifyItemRangeChanged for our case;
        setSelections(false)
    }

    private fun setSelections(isSelected: Boolean) {
        for (binding in bindings) {
            binding.checkbox.isChecked = isSelected
        }
    }

    private val searchFilter: Filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence?): FilterResults? {
            val filteredList = ArrayList<ApplicationItem>()

            if (constraint == null || constraint.isEmpty()) {
                isFiltering = false
                filteredList.addAll(allApps)
            } else {
                isFiltering = true
                val filterPattern = constraint.toString().lowercase(Locale.getDefault()).trim { it <= ' ' }
                for (app in allApps) {
                    if (app.applicationName.lowercase(Locale.getDefault()).contains(filterPattern)) {
                        filteredList.add(app)
                    }
                }
            }
            val results = FilterResults()
            results.values = filteredList
            return results
        }

        override fun publishResults(constraint: CharSequence?, results: FilterResults) {
            if (results.values is List<*>) {
                val oldList = filteredApps
                filteredApps = prepareDataToShow(results.values as ArrayList<ApplicationItem>)
                notifyChanges(oldList, filteredApps)
            }
        }
    }

    fun notifyChanges(oldList: List<SplitTunnelingItem>, newList: List<SplitTunnelingItem>) {
        val diff = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return oldList[oldItemPosition] == newList[newItemPosition]
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return oldList[oldItemPosition] == newList[newItemPosition]
            }

            override fun getOldListSize() = oldList.size
            override fun getNewListSize() = newList.size
        })
        diff.dispatchUpdatesTo(this)
    }

    private fun prepareDataToShow(apps: ArrayList<ApplicationItem>): ArrayList<SplitTunnelingItem> {
        val listToShow = ArrayList<SplitTunnelingItem>()

        listToShow.add(SearchItem())
        listToShow.addAll(apps)

        return listToShow
    }

    internal inner class DescriptionViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
    inner class MenuHandler {
        fun selectAll() {
            this@SplitTunnelingRecyclerViewAdapter.selectAll()
        }

        fun deselectAll() {
            this@SplitTunnelingRecyclerViewAdapter.deselectAll()
        }
    }

    companion object {
        private const val APP_ITEM = 0
        private const val SEARCH_ITEM = 1
    }
}